pyscg-0037: Tweak heading test about assertions (#1083)
ossf/wg-best-practices-os-developers: docs/Secure-Coding-Guide-for-Python/08_coding_standards/pyscg-0037/README.md

> # pyscg-0037: Do not Depend on Assertions In Production

Author: ```suggestion
# pyscg-0037: Presume Assertions May Be Disabled In Production
```

> Assertions are a useful developer tool, but they cannot be relied upon to be present in a production environment. Assertions can be used, but incorrect function arguments should be handled by an appropriate exception instead.

Reviewer: @david-a-wheeler 

I believe that basically 'allowing' assertion code to be present in production code requires to change both sentences and be very clear on when it is safe to have assertions present including reference's supporting that statement.

Author: Forbidding the use of assertions means you have to re-create them, which would be worse. Fuzzers are *much* more effective, for example, when you have assertions, because assertions make it possible for fuzzers to detect more problems.

You *want* assertions in the code, because they make testing more productive. However, since assertions are often disabled in production Python code, the code must work properly when assertions are removed.

Reviewer 2: Hi @david-a-wheeler.
I wonder if we have different meanings of the word "production".

Our understanding was something like this:
Dev = your laptop while coding
Staging = a private preview before launch
Production = the public website everyone visits

We would not say that unit tests are expected to be in a well designed production environment. Perhaps if there's different interpretations of production, we have to do a better job clarifying this in the rule.

I would argue that we are not completely restricting the use of assertions in the rule, as we are saying "Avoid" which basically says "consider something else".

> since assertions are often disabled in production Python code, the code must work properly when assertions are removed.
 
Our understanding is that test code is not supposed to be in production code / live systems.

I would also argue that even if there is usecases in which we want assertions to be in production code, it would be contradictory to what is in the rule at the moment as optimising it via -O/-OO would completely break the code.

Also, generally speaking using -O/-OO to switch between staging and live production is generally a bad idea.

Don't mean to be a pain about this, apologies if that's how it's coming across that way!

Reviewer 3: @david-a-wheeler @BartKaras1128 

Personally, I have not come across any uses of assertions outside of unit tests, so I did my best to check the literature on the topic.

Perhaps we should use the term "public release" (or just "release") instead of "production" to avoid confusion.

"Programming in Python 3" (ISBN-13: 978-0-321-68056-3, p.184-185) does support the idea of optimising the execution via the -O/-OO flags, or even creating a separate production code with the assertions commented out.

"Clean Code in Python" (ISBN 978-1-80056-021-5, ch. "Using assertions in Python", p. 88-90) stresses assertions must not be used when evaluating business logic, but it does suggest they can be used to perform sanity checks and verifying invariants to make sure our algorithms themselves are correct. It even includes the following tip, which discourages from using the -O flag in production:

> Do not run your production programs with python –O … as 
you'd like to take advantage of the assertions in the code to correct 
defects.

It is difficult to find the consensus on the matter but it appears as though assertions in production code are permissable as long as optimisation isn't essential and they aren't used to evaluate business logic. If we want to prevent a scenario where an assumed impossible condition occurs and we have no way of recovery, even a crash using assertions might be an acceptable solution.

Author: > it appears as though assertions in production code are permissable as long as optimisation isn't essential and they aren't used to evaluate business logic

Exactly. Assertions within production code are a powerful and useful technique, which is why practically all programming languages include them. But like all constructs, they have to be used *correctly* - if you use asserts or any other construct incorrectly, it's a bad idea.

> Assertions are a useful developer tool, but they cannot be relied upon to be present in a production environment. Assertions can be used, but incorrect function arguments should be handled by an appropriate exception instead.

Author: ```suggestion
Assertions are a useful developer tool, but they cannot be relied upon to be present in a production environment. Only use assertions when their removal in production is acceptable. In particular, to report invalid input or incorrect function arguments, raise an appropriate exception or other such actions instead of assertions, to ensure that they will continue to occur even when assertions are removed in production.
```

> Assertions are a useful developer tool, but they cannot be relied upon to be present in a production environment. Only use assertions when their removal in production is acceptable. In particular, to report invalid input or incorrect function arguments, raise an appropriate exception or other such actions instead of assertions, to ensure that they will continue to occur even when assertions are removed in production.

Author: ```suggestion
Assertions are a useful developer tool, but they cannot be relied upon to be present in a production environment. Only use assertions when their removal in production is acceptable. In particular, to report invalid input or incorrect function arguments, raise an appropriate exception or other such actions instead of assertions, to ensure that they will continue to occur even when assertions are disabled in production.
```

> Assertions are a useful developer tool, but they cannot be relied upon to be present in a production environment. Only use assertions when their removal in production is acceptable. In particular, to report invalid input or incorrect function arguments, raise an appropriate exception or other such actions instead of assertions, to ensure that they will continue to occur even when assertions are disabled in production.

Author: ```suggestion
Assertions are a useful developer tool, but they cannot be relied upon to be enabled in a production environment. Only use assertions when disabling them in production is acceptable. In particular, to report invalid input or incorrect function arguments, raise an appropriate exception or other such actions instead of using assertions, to ensure that they will continue to occur even when assertions are disabled in production.
```

> Python removes assertions when a script is run with the `-O`  and `-OO` options [[Python 3.9 Documentation](https://docs.python.org/3.9/using/cmdline.html?highlight=pythonoptimize#cmdoption-o)]. The `-O` options is for optimisation. It removes asserts statements from bytecode, removes docstrings from functions/classes and sets `__debug__` to False. It is used for slightly faster execution and smaller bytecode files. `-OO` does everything that `-O`does but it additionally removes module-level docstrings and creates an even more compact bytecode.

Reviewer: @david-a-wheeler after closer inspection of assert I believe now that you are correct about assert statements. 

A blanket prohibition of `assert` is not appropriate and would lead to false positives, as it is legitimately used in the Python standard library.
see `egrep -Ru assert /usr/lib/python3.9/ 2>/dev/null|egrep -v "unittest|test_support|test"`.

The issue I see now is that its not about *where* but *how* `assert` is used. It looks like the discussion got a bit derailed, apologies for that. 

```suggestion
suggestion
Only use `assert` for internal invariants and debug-only checks that are *not required for correct or safe execution*.

__Do not use `assert` for:__
* Security checks (including input validation)
* Enforcing runtime conditions whose removal would affect correctness or security
* Error handling

__Reason:__

Python’s optimization options (`-O` and `-OO`) remove `assert` statements from the bytecode and set `__debug__` to `False`  [[Python 3.9 Documentation](https://docs.python.org/3.9/using/cmdline.html?highlight=pythonoptimize#cmdoption-o)]. This means any logic implemented using `assert` may not execute in optimized runs.

As a result, using `assert` for security or correctness-critical checks can introduce vulnerabilities or undefined behavior regardless if optimization is enabled. Its neither recommended to depend on `-O` or `-OO` as a "switch" to make assertions unreachable. 

Instead, raise appropriate exceptions to ensure the checks are always enforced regardless of interpreter settings.
```

Current title is "pyscg-0037: Avoid Assertions In Production"

We have to remove 'production environment' from our current title, suggested new title: 

@david-a-wheeler : We have been discussing the titles today, came up with more variations and voted:

A) Restrict assert to Internal Non-Critical Checks 3/5
counter argument here was that "Non-Critical Checks" sounds to vague. Also one would only have asserts for use-full and required internal invariant checks. 
 
B) Do not Depend on Assertions in Production 3/5
Counter argument, still holds onto "production".

C) Use Assertions Only for Internal Invariants and Debug-Only Checks 5/5
While it does miss out on "reachable" it does capture the only reasons for us to leave assert in production code. The title would get to long tho, so capping it to only the essence. 

Rg Helge

# pyscg-0037: Presume Assertions May Be Disabled In Production

> suggestion

Author: ```suggestion
Assertions are a useful developer tool, but they cannot be relied upon to be applied in a production environment. 
```

> Assertions are a useful developer tool, but they cannot be relied upon to be applied in a production environment. 

Author: ```suggestion
Assertions are a useful developer tool, but they cannot be relied upon to be applied in a production environment.
```

Only use `assert` for internal invariants and debug-only checks that are *not required for correct or safe execution*.

__Do not use `assert` for:__

> * Security checks (including input validation)

Author: ```suggestion

* Security checks (including input validation)
```

* Enforcing runtime conditions whose removal would affect correctness or security
* Error handling

__Reason:__

Python’s optimization options (`-O` and `-OO`) remove `assert` statements from the bytecode and set `__debug__` to `False`  [[Python 3.9 Documentation](https://docs.python.org/3.9/using/cmdline.html?highlight=pythonoptimize#cmdoption-o)]. This means any logic implemented using `assert` may not execute in optimized runs.

> As a result, using `assert` for security or correctness-critical checks can introduce vulnerabilities or undefined behavior regardless if optimization is enabled. Its neither recommended to depend on `-O` or `-OO` as a "switch" to make assertions unreachable. 

Author: ```suggestion
As a result, using `assert` incorrectly (that is, for security or correctness-critical checks) can introduce vulnerabilities or other undesired behavior when optimization is enabled.
```

It won't be *undefined* behavior in Python, that's a C-ism.

> # pyscg-0037: Presume Assertions May Be Disabled In Production

Reviewer 4: If we are following APA style then 'in' should be lowercase:
https://apastyle.apa.org/style-grammar-guidelines/capitalization/title-case

Reviewer 4: I wish we weren't following APA style, but I lost that argument :-)

> Assertions are a useful developer tool, but they cannot be relied upon to be applied in a production environment.

Reviewer 4: This sentence is slightly off:

they cannot be relied upon to be applied in a production environment.

“Applied” is not quite the right verb. The Python docs are more precise: with optimization, __debug__ becomes False and the code generator emits no code for assert. So the issue is that assertions may not execute at all, not that they are not “applied.”

I would rewrite it as:

Assertions are a useful developer tool, but production code must not rely on them executing.
